src/credentials/GroupDataProvider.h

```diff
@@ -0,0 +1,236 @@
+        uint8_t num_keys_used;
+    };
+
+    // Iterator for group mappings under a given endpoint. Associated with
+    // Groups cluster logic.
+    class GroupMappingIterator
+    {
+    public:
+        virtual ~GroupMappingIterator() = default;
+        // Returns the number of entries in total that will be iterated.
+        virtual uint16_t Count() = 0;
+        // Returns true if a groupID is found in the iteration.
+        virtual bool Next(GroupId & outGroup) = 0;
+        // Release the memory allocated by this iterator, if any. Must be called before
+        // losing scope of a `GroupMappingIterator *`
+        virtual void Release() = 0;
+
+    protected:
+        GroupMappingIterator() = default;
+    };
+
+    // Iterator for group state information mapping a Group ID to a Group Key Set index,
+    // such as reflected by the Groups attribute of the Group Key Management cluster.
+    class GroupStateIterator
+    {
+    public:
+        virtual ~GroupStateIterator() = default;
+        // Returns the number of entries in total that will be iterated.
+        virtual uint16_t Count() = 0;
+        // Returns true if a GroupState is found in the iteration.
+        virtual bool Next(GroupState & outEntry) = 0;
+        // Release the memory allocated by this iterator, if any. Must be called before
+        // losing scope of a `GroupStateIterator *`
+        virtual void Release() = 0;
+
+    protected:
+        GroupStateIterator() = default;
+    };
+
+    // Iterator for the Group Key Sets related under a given Fabric.
+    // TODO: Refactor to allow trial decryption and encryption directly, rather than accessing raw keys.
+    class KeySetIterator
+    {
+    public:
+        virtual ~KeySetIterator() = default;
+        // Returns the number of entries in total that will be iterated.
+        virtual uint16_t Count() = 0;
+        // Returns true if a KeySet is found in the iteration.
+        virtual bool Next(KeySet & outSet) = 0;
+        // Release the memory allocated by this iterator, if any. Must be called before
+        // losing scope of a `KeySetIterator *`
+        virtual void Release() = 0;
+
+    protected:
+        KeySetIterator() = default;
+    };
+
+    // Interface for a listener for changes in any Group configuration. Necessary
+    // to implement attribute subscription for Group Key Management cluster, and
+    // to react to configuration changes that may impact in-progress functional
+    // work.
+    class GroupListener
+    {
+    public:
+        virtual ~GroupListener() = default;
+        /**
+         *  Listener callback invoked when a GroupState entry is mutated or added.
+         *
+         *  @param[in] old_state  GroupState reflecting the previous entry. Set to nullptr on appends.
+         *  @param[in] new_state  GroupState reflecting the updated/new entry.
+         */
+        virtual void OnGroupStateChanged(const GroupState * old_state, const GroupState * new_state) = 0;
```

Comment:
How will this interface work if the endpoints mapped to a given group change?  GroupState does not provide any mapping to associated endpoints I can see, but `GroupStateStruct` in the spec has an endpoints array.  It seems a write to this attribute could change the associated endpoints.

Reply:
The endpoints attribute is read only in group key management cluster and is meant to reflect what is configured across group cluster instances on the different endpoints

Reply:
OK, thanks.  Makes sense.